### PR TITLE
ffmpeg: bump to master

### PR DESCRIPTION
**Pull request by ReenigneArcher**
## Description

Bump ffmpeg from v5.1 to master (9a3bbf8) and enable new AV1 encoders.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [x] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by psyke83** [comment]
I tested the artifact from the latest run against nightly; `software` and `amfenc` appear to work fine, but I only tested on Windows 11 in SDR.

File: ffmpeg_patches/ffmpeg/05-qsv-session-leak.patch [hunk 1]
```diff
@@ -28,27 +28,21 @@ index 56dffa1f25..504d4f8a45 100644
...
+     AVQSVDeviceContext *hwctx = ctx->hwctx;
+     QSVDevicePriv       *priv = ctx->user_opaque;
  
- static void qsv_device_free(AVHWDeviceContext *ctx)
- {
--    AVQSVDeviceContext *hwctx = ctx->hwctx;
--    QSVDevicePriv       *priv = ctx->user_opaque;
--
 -    if (hwctx->session)
 -        MFXClose(hwctx->session);
```
Comment (cgutman) [inline ffmpeg_patches/ffmpeg/05-qsv-session-leak.patch:36]: I think you want to keep both `MFXClose()` and `MFXUnload()` here

File: ffmpeg_patches/ffmpeg/01-amf-colorspace-v2.patch [hunk 2]
```diff
@@ -323,7 +182,7 @@ index 1ab98d2f78..c5ee9d7e35 100644
 +#include <AMF/components/ColorSpace.h>
  #include <AMF/components/VideoEncoderVCE.h>
  #include <AMF/components/VideoEncoderHEVC.h>
- 
+ #include <AMF/components/VideoEncoderAV1.h>
```
Comment (cgutman) [inline ffmpeg_patches/ffmpeg/01-amf-colorspace-v2.patch:185]: Don't we need to update this patch to set all the color stuff for amfenc_av1?

**Comment by cgutman** [comment]
I think we need to also update `02-idr-on-amf.patch` to cover amf_av1.c

**Comment by psyke83** [comment]
> I think we need to also update `02-idr-on-amf.patch` to cover amfenc_av1.c

I've updated patch 01 (amf hdr) and 05 (qsv) with your suggested changes, but 02 (amf idr insertion) looks more complicated to implement.

The AV1 frame types [seem](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/blob/master/amf/public/include/components/VideoEncoderAV1.h#L103-L110) [different](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/blob/master/amf/public/include/components/VideoEncoderHEVC.h#L89-L96), and I'm not certain how to map the values to the traditional I/B/P/IDR/SKIP frames. The header insertion logic also seems more complicated (with different alignment types), so I'm not 100% certain how to implement the necessary changes.

File: ffmpeg_patches/ffmpeg/05-qsv-session-leak.patch (hunk 1 above)
Comment (cgutman) [reply ffmpeg_patches/ffmpeg/05-qsv-session-leak.patch]: It looks like you removed `MFXUnload()`. We still should have that.

**Review by cgutman — CHANGES_REQUESTED** [review]
We can't merge this because FFmpeg 6.0 introduces a memory leak in the QSV context. We should be able to use FFmpeg 6.1 when it's released.

File: .github/workflows/build-ffmpeg.yml [hunk 3]
```diff
@@ -53,8 +53,8 @@ jobs:
               --enable-amf
               --enable-cuda
               --enable-cuda_llvm
-              --enable-encoder=h264_amf,hevc_amf
-              --enable-encoder=h264_nvenc,hevc_nvenc
+              --enable-encoder=h264_amf,hevc_amf,av1_amf
+              --enable-encoder=h264_nvenc,hevc_nvenc,av1_nvenc
```
Comment (ReenigneArcher) [inline .github/workflows/build-ffmpeg.yml:57]: This should be added to `powerpc64le` as well?

**Comment by cgutman** [comment]
Leaving these notes here so we don't forget:

For FFmpeg 6.1, we'll need to also change `preanalysis` -> `preencode` in our options dict, since upstream FFmpeg changed the semantics of the existing option in https://github.com/FFmpeg/FFmpeg/commit/f593dc98bfe4196ad60510008d43764e526d8cb9

I believe the QSV encoders also gained a `remote_gaming` preset which we should probably use too.

**Comment by psyke83** [comment]
The AMF colourspace patch has also been updated to v3 (which includes av1 support): https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=9415

**Comment by cgutman** [comment]
Testing progress:
- [x] QSV leak fix verified
- [x] Windows AMD H.264+HEVC
- [x] Windows AMD AV1
- [x] Windows Intel H.264+HEVC
- [x] Windows Intel AV1
- [x] Linux VAAPI H.264+HEVC
- [ ] Linux VAAPI AV1
- [ ] Linux NVENC H.264+HEVC
- [ ] Linux NVENC AV1

**Comment by cgutman** [comment]
I'm satisfied that this build generally works. Let's merge this and we can perform final testing on the PR to the Sunshine repo.

**Pull request merged**